doc/specs/#13000 - In-process ConPTY.md

```diff
@@ -0,0 +1,840 @@
+        end
+
+        conhost_parser[VT parser]
+        conhost_buffer[Text Buffer]
+        conhost_renderer[Render Thread]
+        conhost_atlas[Text Renderer]
+
+        server_io-->conhost_parser
+        server_io--VT translation-->VtIo
+        server_io-->VtIo
+        conhost_parser-->VtIo
+        VtIo-->VtEngine
+        conhost_parser-->VtEngine
+        conhost_buffer-->VtEngine
+        conhost_renderer-->VtEngine
+        VtIo-->ConPTY_output
+        VtEngine-->ConPTY_output
+        conhost_buffer--dirty rects-->conhost_renderer
+        conhost_parser-->conhost_renderer
+        conhost_parser--plain text-->conhost_buffer
+        conhost_renderer-->conhost_atlas
+        conhost_buffer--ReadBuffer-->server_io
+    end
+
+    subgraph wt[Windows Terminal]
+        direction LR
+
+        input[ConPTY Input]
+        parser[VT parser]
+        buffer[Text Buffer]
+        renderer[Render Thread]
+        atlas[Text Renderer]
+
+        input-->parser
+        parser-->buffer
+        buffer-->renderer
+        renderer-->atlas
+        parser-->renderer
+    end
+
+    conhost--ConPTY's text pipe across processes-->wt
+
+    linkStyle 2,3,4,5,6,7,9,11 opacity:0.2
+    style VtEngine opacity:0.2
+```
+
+### Goals
+
+* Remove VtEngine
+* Remove `--vtmode`
+* Remove `--resizeQuirk`
+* Add buffering to `VtIo` and hook it up to the output pipe
+* Implement direct VT translations for all relevant Console APIs
+
+### Discussion
+
+The idea is that we can translate Console API calls directly to VT at least as well as the current VtEngine setup can.
+For instance, a call to `SetConsoleCursorPosition` clearly translates directly to a `CUP` escape sequence.
+Effectively, instead of translating output asynchronously in the renderer thread, we'll do it synchronously right during the Console API call.
+
+Apart from the Console APIs, the "cooked read" implementation, which handles our builtin "readline"-like text editor, will need to receive some larger changes as well.
+Its popups use `ReadConsoleOutput` and `WriteConsoleOutput` to backup and restore the affected rectangle.
+It also directly interfaces with the backing text buffer and its translation to VT relies on the existence of VtEngine.
+This results in all of the same issues that were previously outlined.
+In order to solve this, the popups need to be rewritten to use escape sequences so that they can be directly passed to the hosting terminal.
+They should also always be below the current prompt line so that we don't need to perform a potentially lossy backup/restore operation.
+
+The `--vtmode xterm-ascii` switch exists for the telnet client as it only supports ASCII as per RFC854 section "THE NVT PRINTER AND KEYBOARD".
+However, telnet is the only user of this flag and it's trivial to do there (for instance by stripping high codepoints in `WriteOutputToClient` in `telnet/console.cpp`), so there's no reason for us to keep this logic in the new code.
```

Comment:
I'm not entirely sure if I understand how telnet on Windows in particular works, but if you check out `VtConsole.cpp` in the OS repo you can see how it explicitly `CreateProcess`'es `conhost.exe --vtmode ascii ...` and then connects its local pipes to it to then pump the contents somewhere else. This means we can definitely filter the output there if we wanted.

Reply:
I guess that means `VtConsole.cpp` is effectively part of the telnet server? In that case it'll definitely work, because telnet and inbox conpty are updated effectively in lockstep, right?